### ISSUE TITLE
systemd-journald crashed in journal_file_link_data()

### ISSUE THREAD

**Issue by nalimilan** [systemd/systemd#1015]
At least 5 people (including me) experienced this crash in Fedora 20, 21 and 22, with systemd ranging from 208-30 to 219-21.

https://bugzilla.redhat.com/show_bug.cgi?id=1199442

That crash happened at about the same time as a complete system crash. I don't know where's the cause and where's the consequence, nor whether it happened right after rebooting.

```
warning: core file may not match specified executable file.
[New LWP 425]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `/usr/lib/systemd/systemd-journald'.
Program terminated with signal SIGABRT, Aborted.
#0  journal_file_link_data (hash=13263421967165135535, offset=7471400, o=0x7fbc317be128, f=0x7fbc35823d60) at ../src/journal/journal-file.c:781
781 ../src/journal/journal-file.c: No such file or directory.

Thread 1 (Thread 0x7fbc33e40840 (LWP 425)):
#0  journal_file_link_data (hash=13263421967165135535, offset=7471400, o=0x7fbc317be128, f=0x7fbc35823d60) at ../src/journal/journal-file.c:781
        m = 233016
        p = 0
        h = 174279
        r = <optimized out>
#1  journal_file_append_data.lto_priv.143 (f=0x7fbc35823d60, data=0x7fbc3581991d, size=45, ret=0x7fff40f42e48, offset=0x7fff40f42e40) at ../src/journal/journal-file.c:1096
        hash = 13263421967165135535
        p = 7471400
        osize = 109
        o = 0x7fbc317be128
        r = <optimized out>
        compression = <optimized out>
        eq = <optimized out>
        __PRETTY_FUNCTION__ = "journal_file_append_data"
        __func__ = "journal_file_append_data"
#2  0x00007fbc33e9fcf1 in journal_file_append_entry.constprop.104 (f=0x7fbc35823d60, iovec=<optimized out>, n_iovec=24, seqnum=0x7fff40f439f0, offset=0x0, ret=0x0, ts=0x7fff40f42e50) at ../src/journal/journal-file.c:1437
        p = 7275672
        o = 0x7fbc3178e498
        i = <optimized out>
        items = 0x7fff40f42c70
        r = <optimized out>
        xor_hash = 1472606121495878170
        _ts = {realtime = 1440346565536459, monotonic = 156026134}
#3  0x00007fbc33e8bc4b in write_to_journal (priority=39, n=24, iovec=0x7fbc35844130, uid=0, s=0x7fff40f43960) at ../src/journal/journald-server.c:514
        vacuumed = false
        f = 0x7fbc35823d60
        r = <optimized out>
#4  dispatch_message_real.lto_priv.227 (s=0x7fff40f43960, iovec=0x7fbc35844130, n=24, m=66, ucred=<optimized out>, tv=0x7fff40f43720, label=0x7fff40f43760 "system_u:system_r:unconfined_service_t:s0-s0:c0.c1023", label_len=54, unit_id=0x0, priority=39, object_pid=0) at ../src/journal/journald-server.c:821
        pid = "_PID=2722\000e\000\000\000\000\000\025\303"
        uid = "_UID=0\000\000`9\364@\377\177\000\000\000"
        gid = "_GID=0", '\000' <repeats 11 times>, "\063"
        owner_uid = "\337\061\364@\377\177\000\000\276\260\350\063\274\177\000\000_COMM=firefox\000\000"
        source_time = "_SOURCE_REALTIME_TIMESTAMP=1440346565534916\000\377\177\000\000\000"
        o_uid = "_CMDLINE=/usr/lib64/fire"
        o_gid = "\000\000\000\000_EFFECTIVE=0\000\064\364@\377\177\000"
        o_owner_uid = "\320\230\201\065\274\177\000\000\375\233\232\062\274\177\000\000`\t\000\000\000\000\000\000\300\062\364@\377\177\000\000 \004\000\000\000"
        object_uid = 849583517
        object_gid = 0
        x = <optimized out>
        r = <optimized out>
        t = 0x7fbc35847a50 ""
        c = 0x7fbc35862130 ""
        realuid = <optimized out>
        owner = 0
        journal_uid = 0
        owner_valid = false
        audit_session = "fox\000\274\177\000\000\060\066\364@\377\177\000\000P4\364@\377\177\000\000\220\261\350\063"
        audit_loginuid = "8;\316\062\274\177\000\000p;\316\062\274\177\000\000\340\061\364@\377\177\000\000!\316\v\277"
        o_audit_session = "P4\364@\377\177\000\000\345\257\350\063\274\177\000\000\200\060\364@\377\177\000\000\001\000\000\000\377\177\000\000\000\r"
        o_audit_loginuid = "\000\000\000\000\000\000\000\000 ;\316\062\274\177\000\000\060\004\000\000\000\000\000\000\020\000\000\000\000\000\000\000 \004"
        audit = 32700
        loginuid = 897859888
        __PRETTY_FUNCTION__ = "dispatch_message_real"
#5  0x00007fbc33eb35ec in server_dispatch_message (s=<optimized out>, iovec=0x7fbc35844130, n=<optimized out>, m=66, ucred=0x7fff40f43740, tv=0x7fff40f43720, label=0x7fff40f43760 "system_u:system_r:unconfined_service_t:s0-s0:c0.c1023", label_len=54, unit_id=0x0, priority=39, object_pid=0) at ../src/journal/journald-server.c:919
        r = <optimized out>
        path = 0x7fbc35844790 "/system.slice/dbus.service"
        c = <optimized out>
        __PRETTY_FUNCTION__ = "server_dispatch_message"
#6  0x00007fbc33eae301 in server_process_native_message (s=0x7fff40f43960, buffer=<optimized out>, buffer_size=201, ucred=0x7fff40f43740, tv=0x7fff40f43720, label=0x7fff40f43760 "system_u:system_r:unconfined_service_t:s0-s0:c0.c1023", label_len=54) at ../src/journal/journald-native.c:286
        iovec = 0x7fbc35844130
        n = 9
        j = <optimized out>
        tn = 8
        p = 0x7fbc35819999 ""
        remaining = 0
        m = 66
        entry_size = 211
        priority = 39
        identifier = 0x7fbc35848030 "realmd"
        message = 0x7fbc35819560 "quitting realmd service after timeout"
        object_pid = 0
        __PRETTY_FUNCTION__ = "server_process_native_message"
        __func__ = "server_process_native_message"
#7  0x00007fbc33e8fb2a in server_process_datagram (es=<optimized out>, fd=6, revents=<optimized out>, userdata=0x7fff40f43960) at ../src/journal/journald-server.c:1213
        iovec = {iov_base = 0x7fbc358198d0, iov_len = 24575}
        fds = <optimized out>
        n_fds = <optimized out>
        m = <optimized out>
        label_len = <optimized out>
        control = {cmsghdr = {cmsg_len = 32, cmsg_level = 1, cmsg_type = 29, __cmsg_data = 0x7fff40f43720 "\305\361\331U"}, buf = " \000\000\000\000\000\000\000\001\000\000\000\035\000\000\000\305\361\331U\000\000\000\000\204)\b\000\000\000\000\000\034\000\000\000\000\000\000\000\001\000\000\000\002\000\000\000\242\n", '\000' <repeats 14 times>, "F\000\000\000\000\000\000\000\001\000\000\000\003\000\000\000system_u:system_r:unconfined_service_t:s0-s0:c0.c1023", '\000' <repeats 226 times>}
        sa = {sa = {sa_family = 0, sa_data = '\000' <repeats 13 times>}, in = {sin_family = 0, sin_port = 0, sin_addr = {s_addr = 0}, sin_zero = "\000\000\000\000\000\000\000"}, in6 = {sin6_family = 0, sin6_port = 0, sin6_flowinfo = 0, sin6_addr = {__in6_u = {__u6_addr8 = '\000' <repeats 15 times>, __u6_addr16 = {0, 0, 0, 0, 0, 0, 0, 0}, __u6_addr32 = {0, 0, 0, 0}}}, sin6_scope_id = 0}, un = {sun_family = 0, sun_path = '\000' <repeats 107 times>}, nl = {nl_family = 0, nl_pad = 0, nl_pid = 0, nl_groups = 0}, storage = {ss_family = 0, __ss_align = 0, __ss_padding = '\000' <repeats 111 times>}, ll = {sll_family = 0, sll_protocol = 0, sll_ifindex = 0, sll_hatype = 0, sll_pkttype = 0 '\000', sll_halen = 0 '\000', sll_addr = "\000\000\000\000\000\000\000"}}
        n = <optimized out>
        ucred = <optimized out>
        tv = <optimized out>
        cmsg = <optimized out>
        label = <optimized out>
        msghdr = {msg_name = 0x7fff40f43600, msg_namelen = 0, msg_iov = 0x7fff40f435b0, msg_iovlen = 1, msg_control = 0x7fff40f43710, msg_controllen = 136, msg_flags = 1073741824}
        v = 201
        s = <optimized out>
        __PRETTY_FUNCTION__ = "server_process_datagram"
        __func__ = "server_process_datagram"
#8  0x00007fbc33ea48a0 in source_dispatch (s=s@entry=0x7fbc35818570) at ../src/libsystemd/sd-event/sd-event.c:2115
        r = <optimized out>
        __PRETTY_FUNCTION__ = "source_dispatch"
        __func__ = "source_dispatch"
#9  0x00007fbc33ea5450 in sd_event_dispatch (e=0x7fbc35818190) at ../src/libsystemd/sd-event/sd-event.c:2472
        p = <optimized out>
        r = <optimized out>
        __PRETTY_FUNCTION__ = "sd_event_dispatch"
#10 0x00007fbc33e86d6e in sd_event_run (timeout=18446744073709551615, e=0x7fbc35818190) at ../src/libsystemd/sd-event/sd-event.c:2501
        r = <optimized out>
#11 main (argc=<optimized out>, argv=<optimized out>) at ../src/journal/journald.c:109
        t = 18446744073709551615
        n = <optimized out>
        server = {syslog_fd = 4, native_fd = 6, stdout_fd = 5, dev_kmsg_fd = 10, audit_fd = 3, hostname_fd = 11, event = 0x7fbc35818190, syslog_event_source = 0x7fbc35818480, native_event_source = 0x7fbc35818570, stdout_event_source = 0x7fbc35818390, dev_kmsg_event_source = 0x7fbc35818660, audit_event_source = 0x7fbc35818750, sync_event_source = 0x7fbc3581fe80, sigusr1_event_source = 0x7fbc35818b40, sigusr2_event_source = 0x7fbc35818c30, sigterm_event_source = 0x7fbc35818d20, sigint_event_source = 0x7fbc35818e10, hostname_event_source = 0x7fbc35818840, runtime_journal = 0x0, system_journal = 0x7fbc35823d60, user_journals = 0x7fbc35817190, seqnum = 4640, buffer = 0x7fbc358198d0 "CODE_FILE=service/realm-daemon.c\nCODE_LINE=420\nCODE_FUNC=realm_daemon_syslog\nMESSAGE=quitting realmd service after timeout\nREALMD_OPERATION=(null)\nPRIORITY=7\nSYSLOG_FACILITY=4\nSYSLOG_IDENTIFIER=realmd"..., buffer_size = 24576, rate_limit = 0x7fbc35818f40, sync_interval_usec = 300000000, rate_limit_interval = 30000000, rate_limit_burst = 1000, runtime_metrics = {max_use = 295620608, use = 18446744073709551615, max_size = 36954112, min_size = 4194304, keep_free = 443428864}, system_metrics = {max_use = 4022489088, use = 18446744073709551615, max_size = 134217728, min_size = 4194304, keep_free = 4294967296}, compress = true, seal = true, forward_to_kmsg = false, forward_to_syslog = false, forward_to_console = false, forward_to_wall = true, n_forward_syslog_missed = 0, last_warn_forward_syslog_missed = 0, cached_available_space = 0, cached_available_space_timestamp = 130112851, var_available_timestamp = 0, max_retention_usec = 0, max_file_usec = 2629800000000, oldest_file_usec = 1398428087495774, stdout_streams = 0x7fbc35831210, n_stdout_streams = 104, tty_path = 0x0, max_level_store = 7, max_level_syslog = 7, max_level_kmsg = 5, max_level_console = 6, max_level_wall = 0, storage = STORAGE_AUTO, split_mode = SPLIT_UID, mmap = 0x7fbc35816060, dev_kmsg_readable = true, kernel_seqnum = 0x7fbc33e7b000, udev = 0x7fbc35818f20, sync_scheduled = true, machine_id_field = "_MACHINE_ID=f6bd7478243382f240fc43fd6bbc94b1", boot_id_field = "_BOOT_ID=19957ee7c24c48dd946aa5f897e018eb", hostname_field = 0x7fbc358193a0 "_HOSTNAME=milan", cgroup_root = 0x7fbc358195b0 "/"}
        r = <optimized out>
        __func__ = "main"
```

**Comment by poettering** [comment]
What systemd version is that backtrace for?

**Comment by nalimilan** [comment]
Ah, indeed. That's with 219-21 on Fedora 22.

**Comment by poettering** [comment]
v223 had a bunch of fixes for the journal (see commit dade37d403f1b8c1d7bb2efbe2361f2a3e999613 and around). Please try to reproduce the issue with a version that includes this commit (and the ones around it) first, and provide a backtrace for that version, then reopen. Thanks!

**Comment by nalimilan** [comment]
OK, thanks, though I'll need to wait for it to enter Fedora first (I really don't have time for this at the moment, I just wanted to get my system stop crashing :-).

Any idea about whether it could be responsible for a system crash at all?
